documentation/man/man3/GameMaster.3
.TH "GameMaster" 3 "Samedi 30 Juillet 2022" "Jeu POA" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GameMaster
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <GameMaster\&.hpp>\fP
.SS "Fonctions membres publiques"

.in +1c
.ti -1c
.RI "\fBGameMaster\fP ()"
.br
.ti -1c
.RI "\fB~GameMaster\fP ()"
.br
.ti -1c
.RI "void \fBendApplication\fP ()"
.br
.ti -1c
.RI "void \fBupdateDeltatime\fP ()"
.br
.ti -1c
.RI "void \fBupdateSFMLEvents\fP ()"
.br
.ti -1c
.RI "void \fBupdate\fP ()"
.br
.ti -1c
.RI "void \fBrender\fP ()"
.br
.ti -1c
.RI "void \fBrun\fP ()"
.br
.in -1c
.SS "Fonctions membres publiques statiques"

.in +1c
.ti -1c
.RI "static const \fBGameMaster\fP * \fBGAME_MASTER\fP ()"
.br
.ti -1c
.RI "static \fBStateData\fP * \fBSTATE_DATA\fP ()"
.br
.ti -1c
.RI "static const sf::Vector2i \fBgetSPRITE_BOX_CENTER\fP ()"
.br
.ti -1c
.RI "static float \fBgetSCREEN_SCALING_FACTOR\fP ()"
.br
.in -1c
.SH "Description détaillée"
.PP 
Définition à la ligne \fB16\fP du fichier \fBGameMaster\&.hpp\fP\&.
.SH "Documentation des constructeurs et destructeur"
.PP 
.SS "GameMaster::GameMaster ()"

.PP
Définition à la ligne \fB145\fP du fichier \fBGameMaster\&.cpp\fP\&.
.SS "GameMaster::~GameMaster ()"

.PP
Définition à la ligne \fB162\fP du fichier \fBGameMaster\&.cpp\fP\&.
.SH "Documentation des fonctions membres"
.PP 
.SS "void GameMaster::endApplication ()"

.PP
Définition à la ligne \fB199\fP du fichier \fBGameMaster\&.cpp\fP\&.
.SS "const \fBGameMaster\fP * GameMaster::GAME_MASTER ()\fC [static]\fP"

.PP
Définition à la ligne \fB183\fP du fichier \fBGameMaster\&.cpp\fP\&.
.SS "float GameMaster::getSCREEN_SCALING_FACTOR ()\fC [static]\fP"

.PP
Définition à la ligne \fB193\fP du fichier \fBGameMaster\&.cpp\fP\&.
.SS "const sf::Vector2i GameMaster::getSPRITE_BOX_CENTER ()\fC [static]\fP"

.PP
Définition à la ligne \fB178\fP du fichier \fBGameMaster\&.cpp\fP\&.
.SS "void GameMaster::render ()"

.PP
Définition à la ligne \fB257\fP du fichier \fBGameMaster\&.cpp\fP\&.
.SS "void GameMaster::run ()"

.PP
Définition à la ligne \fB289\fP du fichier \fBGameMaster\&.cpp\fP\&.
.SS "\fBStateData\fP * GameMaster::STATE_DATA ()\fC [static]\fP"

.PP
Définition à la ligne \fB188\fP du fichier \fBGameMaster\&.cpp\fP\&.
.SS "void GameMaster::update ()"

.PP
Définition à la ligne \fB233\fP du fichier \fBGameMaster\&.cpp\fP\&.
.SS "void GameMaster::updateDeltatime ()"

.PP
Définition à la ligne \fB204\fP du fichier \fBGameMaster\&.cpp\fP\&.
.SS "void GameMaster::updateSFMLEvents ()"

.PP
Définition à la ligne \fB211\fP du fichier \fBGameMaster\&.cpp\fP\&.

.SH "Auteur"
.PP 
Généré automatiquement par Doxygen pour Jeu POA à partir du code source\&.


src/include/GameMaster.cpp
#include "include/GameMaster.hpp"
#include "include/Collider.hpp"
#include "include/MainMenuState.hpp"
#include "include/Position.hpp"

class Collider;
class MainMenuState;
class Position;

using namespace std;

//
const sf::Vector2i GameMaster::m_spriteBoxCenter(17, 16);
int GameMaster::m_countInstance = 0;
GameMaster* GameMaster::m_gameMaster = nullptr;
float GameMaster::screenScalingFactor(1.f);
StateData* GameMaster::m_static_stateData = nullptr;

//
void GameMaster::initGraphicsSettings()
{
	// NOTE - Because we use this->windowrender; we must parialy init it here...
	this->window = new sf::RenderWindow();

	//
	this->graphicsSettings.loadFromFile("Config/window.ini");

	util::Platform platform;
	// in Windows at least, this must be called before creating the window
	float screenScalingFactor = platform.getScreenScalingFactor(this->window->getSystemHandle());
	GameMaster::screenScalingFactor = screenScalingFactor;
	// game::SCREEN_SCALING_FACTOR = screenScalingFactor;

	// Update the graphicsSetting.resolution with screenScalingFactor
	// this->graphicsSettings.resolution.width *= screenScalingFactor;
	// this->graphicsSettings.resolution.height *= screenScalingFactor;
	// this->graphicsSettings.setScreenScalingFactor(screenScalingFactor);

	// We set the icon here...
	platform.setIcon(window->getSystemHandle());

	// // Initialisation du vidéo mode
	// game::vm = &this->graphicsSettings.resolution;
}

void GameMaster::initWindow()
{
	sf::VideoMode vm = this->graphicsSettings.resolution;
	// // Use the screenScalingFactor
	// vm.width *= screenScalingFactor;
	// vm.height *= screenScalingFactor;

	// Create the main window
	if (this->graphicsSettings.fullscreen)
	{
		window->create(vm, this->graphicsSettings.title, sf::Style::Fullscreen);
		// this->window = new sf::RenderWindow(
		// 	vm,
		// 	// this->graphicsSettings.resolution, //? vm à la place
		// 	this->graphicsSettings.title,
		// 	sf::Style::Fullscreen /*,
		// 	this->graphicsSettings.contextSettings*/
		// );
	}
	else
	{

		window->create(vm, this->graphicsSettings.title, sf::Style::Close);
		// this->window = new sf::RenderWindow(
		// 	vm,
		// 	// this->graphicsSettings.resolution, //? vm à la place
		// 	this->graphicsSettings.title,
		// 	sf::Style::Titlebar | sf::Style::Close /*,
		// 	this->graphicsSettings.contextSettings*/
		// );
	}

	this->window->setFramerateLimit(this->graphicsSettings.frameRateLimit);
	this->window->setVerticalSyncEnabled(this->graphicsSettings.verticalSync);

	// Initialisations de l'espace pour tous les éléments du jeux
	Position::initSpace(game::GAME_MAP_WIDTH, game::GAME_MAP_HEIGHT);

	//! Construction de la maison
	// House house;
	// if (!house.load("content/house.png", sf::Vector2u(32, 32), house.getDisposition(), 40, 80))
	// {
	// 	std::cerr << "Erreur chargement < content/Tiles.png >" << std::endl;
	// 	return exit(-1);
	// }

	/* TODO - C'est pour la partie GameState
	// Mise en place des vues
	// sf::View player_view(sf::Vector2f(350.f, 300.f), sf::Vector2f(1000.f, 600.f));
	sf::View player_view;
	// player_view.setCenter(sf::Vector2f(WINDOW_WIDTH / 2.f, WINDOW_HEIGHT / 2.f));
	player_view.setCenter(sf::Vector2f(32.f * 4.f, 32.f * 78.f));
	player_view.setSize(sf::Vector2f(GameMap::getGAME_MAP_WIDTH() / 2, GameMap::getGAME_MAP_HEIGHT() / 8));
	sf::View minimap_view;
	minimap_view.setViewport(sf::FloatRect(0.75f, 0.f, 0.25f, 0.25f));

	// activation de la vue
	window->setView(player_view);
	// window->setView(minimap_view);
	//*/

	// Personage* spawin = new Soldier(true);
	// // Personage spawin = Personage(true);
	// spawin->setGameObjectName("spawin");

	// Personage* p2 = new Soldier();

	// auto chrono = sf::Clock();
	// sf::Event event;
}

void GameMaster::intiStateData()
{
	this->stateData.window = this->window;
	this->stateData.graphicsSettings = &this->graphicsSettings;
	// this->stateData.supportedKeys = &this->supportedKeys;
	this->stateData.states = &this->states;
	// this->stateData.gridSize = this->gridSize;

	this->initFonts();

	GameMaster::m_static_stateData = &this->stateData;
}

void GameMaster::intiStates()
{
	this->states.push(new MainMenuState(&this->stateData));
	// this->states.push(new GameState(this->window));
}

void GameMaster::initFonts()
{
	if (!this->stateData.defaultFont.loadFromFile("content/fonts/Dosis-Light.ttf"))
	{
		throw("ERROR::GAMEMASTER::COULD_NOT_LOAD_FONT");
	}
}
// ---------------------------------------

GameMaster::GameMaster(/*util::Platform& platform*/)
{
	m_countInstance++;
	if (m_countInstance > 1)
	{
		cerr << "LE game master ne doit pas etre instancié plus d'une fois" << endl;
		exit(-1);
	}

	this->initGraphicsSettings();
	this->initWindow();
	this->intiStateData();
	this->intiStates();

	GameMaster::m_gameMaster = this;
}

GameMaster::~GameMaster()
{
	delete this->window;
	// delete spawin;
	// delete p2;

	// REVIEW -
	while (!this->states.empty())
	{
		delete this->states.top();
		this->states.pop();
	}
}

// ------------------------------------

const sf::Vector2i GameMaster::getSPRITE_BOX_CENTER()
{
	return m_spriteBoxCenter;
}

const GameMaster* GameMaster::GAME_MASTER()
{
	return m_gameMaster;
}

StateData* GameMaster::STATE_DATA()
{
	return m_static_stateData;
}

float GameMaster::getSCREEN_SCALING_FACTOR()
{
	return screenScalingFactor;
}

// -----------------------------------
void GameMaster::endApplication()
{
	cout << "fin application" << endl;
}

void GameMaster::updateDeltatime()
{
	this->deltaTime = this->dtClock.restart().asSeconds();
	// initialisation du time du game object
	GameObject::SetTime(this->deltaTime);
}

void GameMaster::updateSFMLEvents()
{

	while (this->window->pollEvent(this->sfEvent))
	{
		// Close window: exit
		if (this->sfEvent.type == sf::Event::Closed)
			this->window->close();

		if (this->sfEvent.type == sf::Event::Resized)
		{
			// TODO -
		}

		// On envois les events sfml au state en cours
		if (!this->states.empty())
		{
			this->states.top()->updateSFMLEvents(this->sfEvent);
		}
	}
}

void GameMaster::update()
{
	this->updateSFMLEvents();

	// Update items
	if (!this->states.empty())
	{
		this->states.top()->update(this->deltaTime);

		if (this->states.top()->getQuit())
		{
			this->states.top()->endState();
			delete this->states.top();
			this->states.pop();
		}
	}
	// Application end
	else
	{
		this->endApplication();
		this->window->close();
	}
}

void GameMaster::render()
{

	//// initialisation du time du game object
	//// GameObject::SetTime(chrono.restart().asSeconds());

	// Appel pour tester la proximité de chaque collider
	Collider::update();

	//* ANCHOR - Appel des update
	// spawin->update();
	// p2->update();

	// Clear screen
	this->window->clear(); // NOTE -
	// ----------------------
	// Render items
	if (!this->states.empty())
		this->states.top()->render(this->window);

	// ----------------------
	//* Affichage de la maison
	// this->window->draw(house);

	//* ANCHOR - Affichage des gameObjects
	// spawin->show(this->window);
	// p2->show(this->window);

	// Update the window
	this->window->display(); // NOTE -
}

void GameMaster::run()
{

	// Start the game loop
	while (this->window->isOpen())
	{
		this->updateDeltatime();
		this->update();
		// this->window->clear(); // REVIEW - Il n'était pas là et ça marchait
		this->render();
	}
}
